### PR TITLE
PCSX2-WX: Add a dialog for title bar customization

### PR DESCRIPTION
**Pull request by ssakash**
**Summary of changes**:

* Adds a new ``Titlebar customization dialog`` which enables you to remove/display some title bar elements.
* Allows optional display of current video mode on title bar.

image1

Image2

Image3

### CODE REVIEW

**Comment by gregory38** [comment]
It is a good idea :+1:

**Comment by avih** [comment]
Doesn't the BIOS use a more "normal" mode? like NTSC or PAL or whatever? The TV displays it, so it's whatever mode the system defaults to most probably.

**Comment by ssakash** [comment]
> It is a good idea :+1:

Thanks :)

> Doesn't the BIOS use a more "normal" mode? like NTSC or PAL or whatever? The TV displays it, so it's whatever mode the system defaults to most probably.

It uses a different type of NTSC/PAL video mode as mentioned on https://github.com/PCSX2/pcsx2/pull/1491#issuecomment-236576758. Quoting my own post (slightly edited for correction),

```
Sony's binary loader which does regional checkout uses PAL/NTSC DVD video mode (not the regular PAL/NTSC video mode).
```

> Also, the commie messages don't mention it, so I'll ask: does it override the title template? what if there's a conflict between the template which the user sets and the values which this dialog shows?

Currently the options at the dialog has been coded to work only for the default title template, it won't work for user edited templates and it _won't_ cause any conflicts. Maybe it would be a good idea to gray out the check boxes when an user has edited the default title template ?

**Comment by Blackbird88** [comment]
Isn't there also something in ini to configure the EE/VU etc. positions? Will that be exposed too?
Just asking. Not that I **really** need it.

**Comment by avih** [comment]
> uses PAL/NTSC DVD

If that indeed is the case, then that should be the mode IMO. Not "BIOS video mode"...

> Currently the options at the dialog has been coded to work only for the default title template, it won't work for user edited templates and it won't cause any conflicts.

OK, that sounds good. This info should be part of the commit message then.

> Maybe it would be a good idea to gray out the check boxes when an user has edited the default title template ?

Don't know. There might be some solution here that would work reasonably.

However, in general IMO we should not add this feature.

What it does is let people have limited control of the format of the game's window title - which they can do already with more options via the template if they really want to, and which is likely to be invisible during actual gameplay anyway (because most people probably play full screen), and where the value of some of those at the title is likely to get very diminished once we have the OSD.

And for this we're modifying 13 files and adding almost 150 lines of code...

I think the effort (which has been put in already, at least to some degree) and the risk (which always comes with more code) is not worth the added value, especially when you can get more value by just editing the template directly at the ini.

**Comment by avih** [comment]
The problem which this PR tries to solve is is that some people might not know that they can edit the template at the ini or that even if they do, they might not know all the available template "variables" which they can use.

What I argue is that most people really don't mind much or at all as long as what they get by default is useful enough, and it is - all the info is there.

And it's not a problem we're seeing coming up again and again at the forum either. On the contrary, it rarely comes up, if at all. And on the rare cases which it does, we can direct people to edit the template at the ini.

In other words, it's too much (effort, code, risk) to solve a problem which doesn't exist. So overall it has more negative weight than positives.

**Comment by avih** [comment]
Also, I have no problem with adding or improving title template variables. I get that the video mode and/or its vertical frequency or frame rate could be useful in some scenarios to developers, and also that some users might want to use them.

As long as it's done in a sensible way, that would be fine with me, as well as improving the default template in a way which is likely to benefit users.

What I don't think we should do is add GUI to control the template, for the reasons I stated above.

**Comment by ssakash** [comment]
> If that indeed is the case, then that should be the mode IMO. Not "BIOS video mode"...

AFAIK only the binary loader uses those two video modes. ``BIOS`` might sound less confusing to the users compared to ``NTSC/PAL DVD`` and would also avoid having two separate elements on the enum hence I had used it. However I'm open to suggestions.

> Don't know. There might be some solution here that would work reasonably, but IMO it's likely to be sensitive and bug prone.

It's not possible to find an effective solution which replaces elements on user configured templates, they could have set the title template to any random format as possible.

> The problem which this PR tries to solve is that some people might want to change the title but don't know that they can edit the template at the ini or that even if they do, they might not know all the available template "variables" which they can use.

That's not the only issue, there are some others.

* You can't edit the title template at INI file when PCSX2 is running, however the dialog allows such dynamic customizations.
* it's easier to tick some check boxes to disable certain elements on the GUI instead of accessing an INI file frequently.

> What I don't think we should do is add GUI to control the template, for the reasons I stated above.

Well, all you had mentioned was it is a waste of _effort_ , _code_ and  _time_. I don't see any other valid reasons from an user perspective.

**Comment by avih** [comment]
> That's not the only issue, there are some others.
>    You can't edit the title template at INI file when PCSX2 is running, however the dialog allows such dynamic customizations.
>     it's easier to tick some check boxes to disable certain elements on the GUI instead of accessing an INI file frequently.

How frequently it happens to you that you want to change it while the game is running? how frequently users want to change the template while the game is running? it's a solution for a problem which doesn't exist.

Sure, if people wanted to change it every two minutes, then UI would be useful. But the vast vast vast majority of users don't want to change it every two minutes or even care about it at all.

> Well, all you had mentioned was it is a waste of effort , code and time. I don't see any other valid reasons from an user perspective.

No. I mentioned it also adds risk (for bugs), code complexity (harder to maintain), UI complexity, etc. But most importantly, it has very little value for most users, and compared to that, the negatives outweight the value which it does add IMO.

Most projects have guidelines that before a non trivial PR is sent, the subject should be first discussed with other developers to hear if they think it's worth it.

Not every new feature is always a good thing, since every new feature has costs (beyond the initial effort), in terms of maintenance, complexity, risks, etc. And those costs have to be weighted against the value which the feature adds. Sometimes the value a feature adds is worth the costs. Sometimes it does not.

In this case, it's my opinion that it does not (again, only for the UI part. I'm fine with more/improved template variables).

**Comment by Blackbird88** [comment]
Oh my. What have I started.
*hides*

**Comment by avih** [comment]
> Oh my. What have I started.

Not sure where you started it, but it's probably fine.

It's the user's job to inform of issues and come up with requests, and it's the developers' jobs to weight also other factors and decide how they can best help users in a way which balances the needs and costs and usefulness, etc.

And developers are also human being which can make mistakes in judgment or otherwise. Requests and then discussions about them are the key to making good decisions.

**Comment by FlatOutPS2** [comment]
Why not leave the full customization to the ini and add three simple options; minimal, default and extra.
`Minimal` only showing essential fluctuating functions like speed and EE/GS percentages, `Default` showing the normal layout, but without the interlacing option(changes depending on the game, not PCSX2 user setting). And `Extra` showing all including interlacing and the video mode.

**Comment by avih** [comment]
> add three simple options; minimal, default and extra.

Because it would about the same amount of additional code, for even less value.

**Comment by avih** [comment]
On the other hand, if one wants to add a dialog to edit the title (in text) where the available variables are explained, I could live with that. I still think it would be a waste of time and effort for very little value, but if someone insists on adding it I won't object.

**Comment by FlatOutPS2** [comment]
> Because it would about the same amount of additional code, for even less value.

It would be a lot more user friendly, as it doesn't need to open a new seperate window and can work like selecting between ISO's, and it doesn't require editting ini's, which can be daunting for unexperienced users.

**Comment by gregory38** [comment]
```
And for this we're modifying 13 files and adding almost 150 lines of code...

I think the effort (which has been put in already, at least to some degree) and the risk (which always comes with more code) is not worth the added value, especially when you can get more value by just editing the template directly at the ini.
```
Seriously, most files only see a couple of lines change. The `big` change is inside a separate file/class. You can't call that a risk. Otherwise, you can call PCSX2 frozen if you don't allow those small changes. You can't either said that the solution is to edit ini because we could nearly remove the full GUI.

**Comment by avih** [comment]
> You can't either said that the solution is to edit ini because we could nearly remove the full GUI.

Not really. You typically want GUI for things which users want to change frequently, or at least that many users would want to change at least once. This is not one of those things - in my opinion. There are many other things that users can't do, some of them not even with the ini, does it mean we should add GUI for every possible thing which users cannot do now?

We should weight it. If it brings a lot of value, sure, if it doesn't bring any value then even if it was 1 line of code then we should not have it. For everything in between, each of us may have different opinions.

And all I said was my opinion, if everyone other than me thinks we should have it, I don't object. I'm just one person and I don't own the truth.

**Comment by gregory38** [comment]
Ini only options are options for developers/testers. If an option is meant to be used by a standard user, it ought to be inside a GUI. Otherwise I don't see the point to have a GUI if you need to fire a text editor to configure your program even if you need it once.

**Comment by ssakash** [comment]
Added a text control option to edit the title template. Automatic Configuration checkboxes will only work for the default title template (will get grayed out for user customized title templates).

dialog

Does this look good ?

**Comment by avih** [comment]
I think that instead of the checkboxes you should just explain what each variable is, and let users edit it however they want. These checkboxes really have limited value IMO. You can't choose where you want it to appear, it only controls some of the variables, and it's not rocket science to edit a line of text, especially if there's also an example just above it.

If anything, there should be a single button instead of all those checkboxes: "Restore default", and the rest is help how to edit it and an example.

In my opinion.

**Comment by FlatOutPS2** [comment]
What if the user enters random garbage into the manual config box?

**Comment by ssakash** [comment]
> What if the user enters random garbage into the manual config box?

Something like this ? ;)

woot

Added Restore defaults button however I still think that the checkboxes should stay, it's more convenient compared to editing the template text.

About adding information related to variables - I wonder whether it's necessary. Users could probably identify the variables based on the values displayed on the default title bar template. so is it really necessary to add them ? not including them might save a bit of dialog space.

**Comment by FlatOutPS2** [comment]
No, something like this: *(@)#%**FJ()@#JF@#*@$$%Q()??Éøã‗ÑJ$Jvnv;"|0(&%

**Comment by ssakash** [comment]
> No, something like this: (@)#%FJ()@#JF@#@$$%Q()??Éøã‗ÑJ$Jvnv;"|0(&%

random

**Comment by ssakash** [comment]
@turtleli 

Could you check whether the code is working fine on Linux ?

**Comment by FlatOutPS2** [comment]
The checkboxes are more convenient, but don't include all options, which makes it useless for all but the optional videomode.  Instead of adding infromation, it would be better to offer a visual clue to which of the title bar entries the checkboxes point to.

**Comment by ssakash** [comment]
> The checkboxes are more convenient, but don't include all options

Only two options are missing, ``Interlaced/Progressive`` indicator and ``Vsync speed/percentage``. I don't understand why anyone would want to remove the latter. I'll try to add an option for the former when I get a free time.

The other stuffs missing are information which are sent by GSdx. (I don't think we can sort it properly)

**Comment by turtleli** [comment]
Do we need another dialog? Would it fit in Emulation Settings -> GS Window?
Does the dialog scale under different DPI settings?
Variable naming seems a bit inconsistent to me btw.

Just to make your life a misery, I'm not keen on your changes and I personally think something like this would be nice for titlebar customisation:
img
(image from https://msdn.microsoft.com/en-us/library/windows/desktop/bb246415(v=vs.85).aspx )

**Comment by ssakash** [comment]
> Do we need another dialog? Would it fit in Emulation Settings -> GS Window?

It should fit in the GS Window but the elements placement style would look awfully inconsistent and I would need to sort all the elements on a dedicated panel to make it look decent. (eventually people will nitpick on the placements and stuffs with comparison to how it was before and we'll never reach a consensus)

so I had gone with a dedicated dialog approach ;)

> Does the dialog scale under different DPI settings?

No clue, I haven't tested it yet but I _think_ it should work. **(I let you handle the testing)** :+1: 

> Variable naming seems a bit inconsistent to me btw.

I'll fix it when I get some free time.

> Just to make your life a misery, I'm not keen on your changes and I personally think something like this would be nice for titlebar customisation:

Personally, I don't see how it's better. It only has one disadvantage : You can't add [custom texts](https://github.com/PCSX2/pcsx2/pull/1509#issuecomment-239193956) into the title template.

Hmm, why are you not keen on the current changes ? not user friendly enough ? However the current change is indeed a step forward compared to the master branch as there was no GUI for the template function itself. Improvements / Innovative ideas could always be added later.

**Comment by turtleli** [comment]
> It should fit in the GS Window but the elements placement style would look awfully inconsistent and I would need to sort all the elements on a dedicated panel to make it look decent. (eventually people will nitpick on the placements and stuffs with comparison to how it was before and we'll never reach a consensus)
> 
> so I had gone with a dedicated dialog approach ;)

It belongs in Emulation Settings -> GS Window IMO. It's the first place I would think of going if I wanted to change the GS Window title bar.

> No clue, I haven't tested it yet but I think it should work. (I let you handle the testing)

I mentioned it because I think it won't work nicely (particularly the text control which you've set to a specific size).

> Personally, I don't see how it's better. It only has one disadvantage : You can't add custom texts into the title template.

IMO it aligns best with what users want and is a good compromise between power and usability.

> Hmm, why are you not keen on the current changes ? not user friendly enough ?

I don't like how two ways of doing the same thing are being added to the GUI. It feels like one way is the "newbie" way, and the other way is the "expert" way, whereas I would prefer having just one way that everyone will use.

> However the current change is indeed a step forward compared to the master branch as there was no GUI for the template function itself. **Improvements / Innovative ideas could always be added later.**

On this, I quote myself:

> (eventually people will nitpick on the placements and stuffs with comparison to how it was before and we'll never reach a consensus)

Whoops, I quote you ;)

**Comment by ssakash** [comment]
> I don't like how two ways of doing the same thing are being added to the GUI. It feels like one way is the "newbie" way, and the other way is the "expert" way, whereas I would prefer having just one way that everyone will use.

Personally I disagree, the idea was to replicate the same functional changes you could do with the INI value so your suggestion doesn't seem ideal. I don't see the problem on having two different ways ? one is for easily toggling the elements (checkbox) and the other one is for manual configuration.

Well, it's not really new for me on how people are opposed to most of my suggested GUI changes :P

Though as pointed out earlier, everyone seems to have different opinion regarding the placement/stuffs, can't really satisfy everyone with a single format. Hence closing it.

**Comment by avih** [comment]
So you're not going to add the video mode as a possible template variable?

**Comment by ssakash** [comment]
> So you're not going to add the video mode as a possible template variable?

It's still useless, how would any user possibly know the template variable for the video mode ? Adding an UI would be the ideal solution. Good luck if you're planning on adding it.

**Comment by avih** [comment]
> It's still useless, how would any user possibly know the template variable for the video mode ? Adding an UI would be the ideal solution

Maybe they don't have to if it's there by default, for instance? Or if we add a menu item or a button "Customize Game window title" which takes them to a web page which explains how?

This is an orthogonal problem to solve. IIRC the title currently doesn't use all the available variables anyway, so adding another one which those who know (or ask how to) can use is not a bad thing IMO.

I'm not saying it should be there by default, but it does sound to me like a useful thing to have, and personally I don't object to modifying the defaults further as long as we think it's best for our users, and as long as anyone can override those defaults via the template - that's why the template was introduced in the first place, as I'm sure you remember. We've improved it in the past, and nothing prevents us from improving it further.

> Good luck if you're planning on adding it.

You started this, and until you're totally done with it or totally lost interest, you can be sure that other devs won't barge into your work and modify your stuff.

Development is not black or white. You have to take into account what other developers think, and users too of course. So far no one objected to adding a variable to the template, so adding it would be a step in the right direction as far as you're concerned IMHO.

Also, you should understand what's the main _problem_ you're trying to solve, and understand that sometimes there's more than one _solution_ to the same problem.

Also, solving one problem at a time instead of combining everything together into one big problem with one big solution which is all or nothing as far as you're concerned - is likely to end up as "nothing" more frequently than "all". 

For instance, this PR tries to address several problems if I understand it correctly (also from the discussion above):
- Users don't even know they can customize the title.
- Even if they know, they can't do it while PCSX2 is running.
- Even if they know, It's hard to customize the title as text (your solution is checkboxes/GUI).
- Having the GS mode at the title is something which some users would like.

Correct, more or less?

Now, which of those problems is really important for you to solve?

I'd think for you it's most important to solve "they don't even know" and "video mode would be useful", right? Or maybe the only problem you really care about is that the video mode would be nice at the title, and all the rest is just to allow users enabling it? I don't know. You should know.

Let's assume we found great solutions for everything, except for the problem of "can't modify it while PCSX2 is running". Would it be useless without it? would you drop everything because other devs don't want your solution to this specific problem? Or would you live with that and still manage to solve 3 out of the 4 problems you wanted to solve, including the two most important ones?

Would you give up on everything because someone thought (hypothetically) that you should change the checkboxes ordering for instance? Or place them at a different panel?

- Understand exactly what problems you're trying to solve.
- Understand exactly which are the most important ones for you - which without them solved you'd really consider the rest useless.
- Understand that there's more than one solution to most problems.
- And solve one problem at a time.

**Comment by ssakash** [comment]
Seriously, I have no interest on GUI changes anymore. You're free to do whatever you want with the title template stuffs.

Thanks for the write-up though, appreciated. :)
